src/org/mirah/builtins/collection_extensions.mirah

```diff
@@ -70,25 +107,70 @@ class CollectionExtensions
       `b`.toString
     end
   end
-
+  
+  # Create a String which consists of the String representation of each element of this Collection, interleaved with the supplied separator.
+  #
+  #   [1,2,3].join("+") => "1+2+3"
   macro def join(separator)
-    b             = gensym
-    list          = gensym
-    entry         = gensym
-    first_element = gensym
+    b     = gensym
+    list  = gensym
+    entry = gensym
+    nonfirst = gensym
     quote do
-      `first_element` = true
       `list` = `@call.target`
       `b`    = StringBuilder.new
+      `nonfirst` = false
       `list`.each do |`entry`|
-        if `first_element`
-          `first_element` = false
-        else
+        if `nonfirst`
           `b`.append(`separator`)
+        else
+          `nonfirst` = true
         end
         `b`.append(`entry`)
       end
       `b`.toString
     end
   end
+    
+#  macro def to_array_complex(basetype) # just for complex types, maybe this implementation could be auto-choosen if the compiler can deduce that the type supplied is indeed a complex type
+#    list  = gensym
+#    quote do
+#      `list` = `@call.target`
+#      `list`.toArray(`basetype`[`list`.size])
+#    end
+#  end
+
+  # Convert this Collection to a Java-style array. You have to supply the base-type of the array, currently.
+  #
+  #   [1,2,3,4].to_array(int)
+  macro def to_array(basetype) # for primitive types and complex types
+    list  = gensym
+    res   = gensym
+    value = gensym
+    index = gensym
+    quote do
+      `list` = `@call.target`
+      `res`  = `basetype`[`list`.size]
+      `list`.each_with_index do |`value`,`index`|
+        `Call.new(quote{`res`},SimpleString.new('[]='),[ quote { `index` } , quote { ` value ` } ],nil)`
+      end
+      `res`
+    end
+  end
+  
+  # Returns first element of the list.
+  # Because Mirah has Ruby syntax but Java semantics, in case the list is empty, nil is not returned, but an instance of java.lang.IndexOutOfBoundsException is raised. 
+  macro def first()
```

Comment:
@baroquebobcat: You wanted #first, #last to return `nil` in case the List is empty. However, what should be the result of `int[0].first` then? It can't be `nil`, as `nil`is not of type `int`. Should #first, #last be undefined for Java-style arrays?